roles/test-utils/mining-device/src/main.rs

```diff
@@ -322,6 +324,17 @@ impl Device {
                 MiningRoutingLogic::None,
             )
             .unwrap();
+            let mut notify_changes_to_mining_thread = self_mutex
+                .safe_lock(|s| s.notify_changes_to_mining_thread.clone())
+                .unwrap();
+            if notify_changes_to_mining_thread.should_send {
+                notify_changes_to_mining_thread
+                    .sender
+                    .send(())
+                    .await
+                    .unwrap();
+                notify_changes_to_mining_thread.should_send = false;
+            };
```

Comment:
currently `Device` is notifying changes to mining threads on **any** message received from upstream, including  `SubmitShareSuccess`

so every time `Device` submits a new share, it restarts all mining threads and start from scratch, which results in shares being submitted repeatedly:
```
2024-07-17T19:52:30.201538Z  INFO mining_device: Found share with nonce: 1185614795, for target: Some(0x00000005b5246dc512185093364f5f8fb78345dd01ec6397d5cd30e1232d5039), with hash: 0x00000004ac66a0b64a11bca5ad95832c12f28f5e3ce88468b750a538a28f51e8
2024-07-17T19:52:30.203086Z  INFO mining_device: SUCCESS SubmitSharesSuccess { channel_id: 6, last_sequence_number: 3, new_submits_accepted_count: 1, new_shares_sum: 0 }
2024-07-17T19:52:42.398146Z  INFO mining_device: Found share with nonce: 1185614795, for target: Some(0x00000005b5246dc512185093364f5f8fb78345dd01ec6397d5cd30e1232d5039), with hash: 0x00000004ac66a0b64a11bca5ad95832c12f28f5e3ce88468b750a538a28f51e8
2024-07-17T19:52:42.399670Z  INFO mining_device: SUCCESS SubmitSharesSuccess { channel_id: 6, last_sequence_number: 4, new_submits_accepted_count: 1, new_shares_sum: 0 }
```

the suggested changes below make sure that notification is only sent when `Device` receives a `NewMiningJob`, `SetNewPrevHash` or `SetTarget` (did I miss any message that could be relevant on this context? @Fi3 please double check)
```suggestion
            if notify_changes_to_mining_thread.should_send
                && (message_type == const_sv2::MESSAGE_TYPE_NEW_MINING_JOB
                    || message_type == const_sv2::MESSAGE_TYPE_SET_NEW_PREV_HASH
                    || message_type == const_sv2::MESSAGE_TYPE_SET_TARGET)
            {
                notify_changes_to_mining_thread
                    .sender
                    .send(())
                    .await
                    .unwrap();
                notify_changes_to_mining_thread.should_send = false;
            };
```